### PR TITLE
added peopleQuestion

### PR DESCRIPTION
**Pull request by iCarrrot**
Added peopleQuestion based on Open Question.
Added some tests.
Repaired Votes.__str__ method (it cause errors while questions ware deleted).

### CODE REVIEW

File: izp/polls/tests.py [hunk 1]
```diff
@@ -208,6 +208,69 @@ def test_future_current_and_past_question(self):
              '<Question: Past question.>']
         )
 
+    def test_basic_peopleQuestion(self):
```
Comment (czapiga) [inline izp/polls/tests.py]: Not consistent naming convention

File: izp/polls/tests.py [hunk 2]
```diff
@@ -208,6 +208,69 @@ def test_future_current_and_past_question(self):
...
+        """
+        Test for detail view of people question
+        """
+        peopleQuestion = PeopleQuestion(question_text="Question?")
+        peopleQuestion.save()
```
Comment (czapiga) [inline izp/polls/tests.py]: question = PeopleQuestion.objects.create(question_text="Question?")

File: izp/polls/tests.py [hunk 3]
```diff
@@ -208,6 +208,69 @@ def test_future_current_and_past_question(self):
...
+        peopleQuestion.choice_set.create(choice_text="dr Grzegorz Świderski")
+        peopleQuestion.choice_set.create(choice_text="dr Grzegorz Świderski")
+        peopleQuestion.choice_set.create(choice_text="dr Grzegorz Świderski")
+        peopleQuestion.choice_set.create(choice_text="dr Grzegorz Świderski")
+        peopleQuestion.choice_set.create(choice_text="dr Grzegorz Świderski")
+        peopleQuestion.choice_set.create(choice_text="dr Grzegorz Świderski")
+        peopleQuestion.choice_set.create(choice_text="dr Grzegorz Świderski")
+        peopleQuestion.choice_set.create(choice_text="dr Grzegorz Świderski")
+        peopleQuestion.choice_set.create(choice_text="dr Grzegorz Świderski")
+        peopleQuestion.choice_set.create(choice_text="dr Grzegorz Świderski")
```
Comment (czapiga) [inline izp/polls/tests.py]: 1. I will merge #15 first :)
2. Remove this test.
3. Hint for you: If you want to test something related to object uniqueness it is usually enough to create 2 the same objects.
4. How this test is related to your task? Clearly you're just checking if it's possible to create more than one choice with the same choice_text. I've sent list of tasks to everyone and I've told YOU that someone is working on that. There was even PR fixing this issue before you created yours.

File: izp/polls/views.py [hunk 4]
```diff
@@ -21,11 +21,16 @@ def detail(request, question_id):
         return render(request, 'polls/detail.html', {
             'question': question, 'error': "Głosowanie nie jest aktywne"})
     try:
-        openQuestion = OpenQuestion.objects.get(pk=question_id)
-    except OpenQuestion.DoesNotExist:
-        return render(request, 'polls/detail.html', {'question': question})
+        peopleQuestion = PeopleQuestion.objects.get(pk=question_id)
```
Comment (czapiga) [inline izp/polls/views.py]: Refactor ```try - except``` into ```if - elif - else ```.

File: izp/polls/tests.py (hunk 3 above)
Comment (iCarrrot) [reply izp/polls/tests.py]: Ad 4. It tests if question is created correct, and if voting is correct.  If somebody is working on uniqueness, I'm pretty sure that he is not working on uniqueness of class which is on my branch. 
Of course I can remove this test, but on the other hand which test should I create? Because this class is very based on OpenQeustion class, so according to your words i don't need to create them. Also answer uniqueness is made by somebody else, so I shouldn't test it. So again, what should I test?

Comment (iCarrrot) [reply izp/polls/tests.py]: Ok, doesn't matter, I don'y really need this test

File: izp/polls/admin.py [hunk 5]
```diff
@@ -27,3 +28,4 @@ class QuestionAdmin(BaseQuestionAdmin):
 admin.site.register(Question, QuestionAdmin)
 admin.site.register(SimpleQuestion, BaseQuestionAdmin)
 admin.site.register(OpenQuestion, QuestionAdmin)
+admin.site.register(PeopleQuestion, BaseQuestionAdmin)
```
Comment (czapiga) [inline izp/polls/admin.py]: ```BaseQuestionAdmin``` -> ```QuestionAdmin```

File: izp/polls/employers.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
+"""
+Employers iterator
+In new file for clear code.
+"""
+
+
+def employers_iterator():
```
Comment (czapiga) [inline izp/polls/employers.py]: What for? I don't understand why would you create some fancy iterator and later use it just like ordinary array/list/dictionary w/e. Please declare employers array/list/dictionary in this file and remove not needed decoration.

File: izp/polls/templates/polls/OpenDetail.html [hunk 7]
```diff
@@ -0,0 +1,45 @@
+{% extends 'polls/base.html' %} 
+{% block content %}
+
+<h1>{{ question.question_text }}</h1>
+<div class="col-sm-3">
+    <form action="{% url 'polls:vote' question.id %}" method="post">
```
Comment (czapiga) [inline izp/polls/templates/polls/OpenDetail.html]: I think it's overkill. You just have to "inject" one field to  basic form. Please refactor it (hint: "extend" detail.html - you will just "import/inject" field for new choice in case of OpenQuestion / PeopleQuestion). Look how it is done in base.html  - injecting content block.

File: izp/polls/tests.py [hunk 8]
```diff
@@ -208,6 +208,59 @@ def test_future_current_and_past_question(self):
              '<Question: Past question.>']
         )
 
+    def test_basic_people_question(self):
+        """
+        Test for detail view of people question
+        """
+        question = PeopleQuestion(question_text="Question?")
```
Comment (czapiga) [inline izp/polls/tests.py]: ```PeopleQuestion.objects.create``` and you won't need to call ```question.save()```

File: izp/polls/tests.py [hunk 9]
```diff
@@ -208,6 +208,59 @@ def test_future_current_and_past_question(self):
...
+        url = reverse('polls:detail', args=(question.id,))
+        response = self.client.get(url)
+        self.assertContains(response, question.question_text)
+        self.assertContains(response, 'dr Grzegorz Świderski')
+        self.assertContains(response, 'dr hab. Jean-Marie de Nivelle')
+        self.assertContains(
+            response, '<input id="1" type="radio" name="choice" value="1">')
+        self.assertContains(
+            response, '<input id="2" type="radio" name="choice" value="2">')
+        self.assertNotContains(
```
Comment (czapiga) [inline izp/polls/tests.py]: Remove

File: izp/polls/tests.py [hunk 10]
```diff
@@ -208,6 +208,59 @@ def test_future_current_and_past_question(self):
...
+        question.choice_set.create(choice_text="dr Grzegorz Świderski")
+        question.choice_set.create(
+            choice_text="dr hab. Jean-Marie de Nivelle")
+        url = reverse('polls:detail', args=(question.id,))
+        response = self.client.get(url)
+        self.assertContains(response, question.question_text)
+        self.assertContains(response, 'dr Grzegorz Świderski')
+        self.assertContains(response, 'dr hab. Jean-Marie de Nivelle')
+        self.assertContains(
+            response, '<input id="1" type="radio" name="choice" value="1">')
```
Comment (czapiga) [inline izp/polls/tests.py]: Change hardcoded values to real ones (extract them from choice object you've created).

File: izp/polls/templates/polls/OpenDetail.html [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+                    {% for e in employers%}
+                    <option value="{{ e }}">{{ e }} </option>
+                    {% endfor %}
+                </select>
+            </datalist>
+            {% endif %}
+
+        </div>
+        
+{% endblock %}
```
Comment (czapiga) [inline izp/polls/templates/polls/OpenDetail.html]: Add new line at the end of file

File: izp/polls/templates/polls/base.html [hunk 12]
```diff
@@ -21,7 +21,7 @@ <h2>Error!</h2>
             <span><h4>{{error}}</h4></span>
          </div>
          {% endif %}
-      
+
```
Comment (czapiga) [inline izp/polls/templates/polls/base.html]: Remove not needed whitespaces

File: izp/polls/views.py [hunk 13]
```diff
@@ -94,9 +101,14 @@ def vote(request, question_id):
...
                 request, 'polls/detail.html',
-                {'question': question,
-                 'error': "Odpowiedź nie istnieje",
-                 'is_open': is_open})
+                {'question': question, 'error': "Odpowiedź nie istnieje"})
+    if not choice and PeopleQuestion.objects.filter(pk=question.pk).exists():
+        try:
+            choice = Choice.objects.get(
+                question__exact=question, choice_text=new_choice)
+        except Exception as e:
```
Comment (czapiga) [inline izp/polls/views.py]: Refactor this code - I don't want to see exceptions. I've already told you that 2 weeks ago.

File: izp/polls/tests.py [hunk 14]
```diff
@@ -148,25 +182,6 @@ def setUp(self):
         open_question.choice_set.create(choice_text="Odp1")
         open_question.choice_set.create(choice_text="Odp2")
 
-    def test_open_question_with_choices(self):
```
Comment (czapiga) [inline izp/polls/tests.py]: Explain why you have removed tests.

File: izp/polls/tests.py [hunk 15]
```diff
@@ -139,6 +140,39 @@ def test_future_current_and_past_question(self):
...
+        """
+        Test for detail view of empty people question
+        """
+        poll = Poll.objects.create()
+        question = PeopleQuestion.objects.create(
+            question_text="Question?", poll=poll)
+        url = reverse('polls:question_detail', args=(question.id,))
+        response = self.client.get(url)
+        self.assertContains(response, question.question_text)
+        # self.assertContains(response, 'new_choice')
```
Comment (czapiga) [inline izp/polls/tests.py]: remove comments!

File: izp/polls/templates/polls/question_detail.html [hunk 16]
```diff
@@ -1,27 +1,32 @@
...
             </label>
-            <input id="new_choice" name="new_choice" type="text" class="form-control" >
-         </div>
-         {% endif %}
-         {% if is_session %}
-         <button class="btn btn-primary">Głosuj</button>
-         {% endif %}
-      </form>
-   </div>
+            <input id="code" name="code" type="password" class="form-control">
```
Comment (czapiga) [inline izp/polls/templates/polls/question_detail.html]: It's just sad. You're really unaware what is happening in project.

File: izp/polls/templates/polls/open_question_detail.html [hunk 17]
```diff
@@ -0,0 +1,27 @@
+{% extends 'polls/detail.html' %} 
+
+{% block chooseAnswr%}
+<label for="radio" id="people_header">Proszę wybrać jedną z dostępnych propozycji, lub podać własną:</label>
```
Comment (valentinebaliuk) [inline izp/polls/templates/polls/open_question_detail.html]: In html 'for' attribute use if and only if you want to bind label to input. 
i don't see in this file input with 'id' attribute "radio"

File: izp/polls/templates/polls/question_detail.html [hunk 18]
```diff
@@ -1,27 +1,32 @@
...
-         {% for choice in question.choice_set.all %}
-       	 <div class="radio">
-            <label><input id="{{ choice.id }}" type="radio" name="choice" value="{{ choice.id }}">
+
+<h1>{{ question.question_text }}</h1>
+<div class="col-sm-3">
+    <form action="{% url 'polls:vote' question.id %}" method="post">
+        {% csrf_token %} 
+        {% block chooseAnswr%}
+        <label for="radio" id="people_header">Proszę wybrać jedną z dostępnych propozycji:</label>
```
Comment (valentinebaliuk) [inline izp/polls/templates/polls/question_detail.html]: here is no input with 'id' attribute equal to "radio" only div with 'class' attribute "radio". It's not possible to bind label to div. 
You can bind label only to input and 'for' attribute value must be equal to input 'id' attribute value.

**Comment by iCarrrot** [comment]
Ok, I hope it's last commit. Now there is almost no difference between OpenQuestion and PeopleQuestion, so PeopleQuestion should work same way.

File: izp/polls/views.py [hunk 19]
```diff
@@ -169,20 +193,38 @@ def vote(request, question_id):
...
         else:
-            return render(
-                request, 'polls/question_detail.html',
-                {'question': question,
-                 'error': "Odpowiedź nie istnieje",
-                 'is_open': is_open,
-                 'is_session': is_session})
-
-    if not choice and is_open:
+            if is_open:
```
Comment (annaola) [inline izp/polls/views.py]: I don't really understand, why do you specify all those cases. Especially, that in this place the don't change anything - you don't even pass employers list and one or two more fields aren't huge load.

File: izp/polls/views.py [hunk 20]
```diff
@@ -169,20 +193,38 @@ def vote(request, question_id):
...
+                     'is_open': is_open,
+                     'is_peopleQ': is_peopleQ,
+                     'is_session': is_session})
+            else:
+                return render(
+                    request, 'polls/detail.html',
+                    {'question': question,
+                     'error': "Odpowiedź nie istnieje",
+                     'is_session': is_session})
+    if not choice and PeopleQuestion.objects.filter(pk=question.pk).exists():
```
Comment (annaola) [inline izp/polls/views.py]: Well, you check it twice - here and below, where you check is_peopleQ variable, it's the same.

File: izp/polls/views.py [hunk 21]
```diff
@@ -169,20 +193,38 @@ def vote(request, question_id):
...
+        if not Choice.objects.filter(
+                question__exact=question,
+                choice_text=new_choice).exists():
+            choice = Choice.objects.create(
+                question=question, choice_text=new_choice)
+        else:
+            choice = Choice.objects.get(
+                question__exact=question, choice_text=new_choice)
+
+    if not choice and (is_open or is_peopleQ):
```
Comment (annaola) [inline izp/polls/views.py]: As you already mentioned, when a question is a people question, it is also an open question, so checking is_open should be enough. And then, we return to the case above, so probably one of them isn't necessary.

File: izp/polls/tests/test_views.py [hunk 22]
```diff
@@ -122,6 +123,37 @@ def test_open_question_without_choices(self):
...
+        poll = Poll.objects.create()
+        people_question = PeopleQuestion.objects.create(
+            poll=poll, question_text="PeopleQuestion")
+        people_question.activate()
+        people_question.choice_set.create(choice_text="Odp1")
+        people_question.choice_set.create(choice_text="Odp2")
+
+    def test_people_question_with_choices(self):
+        '''
+        Test for detail view of open question
```
[the remark this answers is not in the document]
Comment (iCarrrot) [reply izp/polls/tests/test_views.py]: done

File: izp/polls/views.py (hunk 19 above)
Comment (iCarrrot) [reply izp/polls/views.py]: I know why. It was because few months ago there was another template for normal question and another for open question. But in the meantime templates have been changed and here we are.

File: izp/polls/views.py (hunk 20 above)
Comment (iCarrrot) [reply izp/polls/views.py]: Oh, yeah. Auto-merge.

File: izp/polls/views.py [hunk 23]
```diff
@@ -161,6 +184,7 @@ def vote(request, question_id):
                 'error': "Nie można głosować na istniejącą odpowiedź i \
                           jednocześnie proponować nową",
                 'is_open': is_open,
+                'is_peopleQ': is_peopleQ,
```
[the remark this answers is not in the document]
Comment (iCarrrot) [reply izp/polls/views.py]: done

File: izp/polls/employers.py [hunk 24]
```diff
@@ -0,0 +1,16 @@
...
+Employers list
+In new file for clear code.
+"""
+
+
+employers = []
+try:
+    with open("polls/static/data/employers") as f:
+        for line in f:
+            employers.append(line)
```
Comment (sgorawski) [inline izp/polls/employers.py]: This can be done in a more elegant way without iteration:
```python
employers = f.readlines()
```

File: izp/polls/employers.py [hunk 25]
```diff
@@ -0,0 +1,16 @@
...
+except FileNotFoundError:
+
+    print('\033[93m'
+          + "WARNING: File: polls/static/data/employers not found!\n"
+          + '\033[0m')
```
Comment (sgorawski) [inline izp/polls/employers.py]: Wrap it all in a function, like `get_employers_list`, which will return `employers` - it's a lot better than this global variable

File: izp/polls/static/data/employers [hunk 26]
```diff
@@ -0,0 +1,163 @@
...
+dr hab. Krzysztof Topolski
+dr hab. Roman Urban
+dr hab. Roman Wencel
+dr inż. Łukasz Wojakowski
+dr Błażej Wróbel
+dr Jarosław Wróblewski
+dr Anna Wysoczańska-Kula
+dr hab. Janusz Wysoczański
+dr Grzegorz Wyłupek
+dr hab. Jacek Zienkiewicz
```
Comment (sgorawski) [inline izp/polls/static/data/employers]: There should probably be a newline at the end of the file

File: izp/polls/templates/polls/question_detail.html [hunk 27]
```diff
@@ -16,12 +16,23 @@ <h1>{{ question.question_text }}</h1>
...
+                <input id="new_choice" name="new_choice" type="text" class="form-control"  list="employers">
+                <datalist id="employers">
+                    <select>
+                        {% for e in employers %}
+                        <option value="{{ e }}">{{ e }} </option>
+                        {% endfor %}
+                    </select>
+                </datalist>
+            {% else %}
+                <input id="new_choice" name="new_choice" type="text" class="form-control" >
```
Comment (sgorawski) [inline izp/polls/templates/polls/question_detail.html]: Remove space before closing tag `>`

File: izp/polls/templates/polls/question_detail.html [hunk 28]
```diff
@@ -16,12 +16,23 @@ <h1>{{ question.question_text }}</h1>
...
+            {% endif %}
          </div>
          {% endif %}
          {% if is_session %}
          <button class="btn btn-primary">Głosuj</button>
          {% endif %}
       </form>
    </div>
-{% endblock %}
+{% endblock %}
```
Comment (sgorawski) [inline izp/polls/templates/polls/question_detail.html]: Add newline at the end of the file

File: izp/polls/views.py [hunk 29]
```diff
@@ -27,23 +28,32 @@ def question_detail(request, question_id):
     question = get_object_or_404(Question, pk=question_id)
 
     is_open = OpenQuestion.objects.filter(pk=question.pk).exists()
+    is_peopleQ = PeopleQuestion.objects.filter(pk=question.pk).exists()
```
Comment (sgorawski) [inline izp/polls/views.py]: Rename variable to i.e. `is_people_question` - see PEP8 style guide

**Comment by sgorawski** [comment]
Right now the function `get_employers_list` from the _employers.py_ module is reading data from text file every time it's called, which can be quite often - and is unnecessary. I believe it would be better to wrap it up in a Singleton or something like this to read data from text only once. Apart from that, the code seems ok to me.

**Comment by iCarrrot** [comment]
I thing it is better to come back to global var. It's just one array and wrapping it into tons of code is just triumph of form over content for me.

**Comment by iCarrrot** [comment]
@sgorawski let me know which way should I go.

**Comment by sgorawski** [comment]
After a quick research it seems that wrapping the list up in a class as a static variable is the way to go - does not require "tons of code" like the Singleton, however also provides basic encapsulation, unlike the global variable, which can potentially be changed anywhere in code.

Example implementation:
```python
class Employers:
     __list = None

    @classmethod
    def get_list(cls):
        if cls.__list is None:
            # read data from file
            cls.__list = data
        return cls.__list
```

Then in code you'll just have to call `Employers.get_list()`, which will read data from file only if it's the first call. The double underscores hide the variable from scope outside of the class, so it can't be accessed or modified directly, the only way to read it is via the method.

**Comment by sgorawski** [comment]
For me the implementation seems good. The only thing I can point out is that the tests are exactly the same as the ones for `OpenQuestion`, which I know makes sense, but results in code duplication - not an issue though, just something to think about. But anyway the PR is OK for me now.

**Comment by sgorawski** [comment]
The PR should be updated - right now there is a merge conflict.

**Comment by iCarrrot** [comment]
Done

**Comment by czapiga** [comment]
Still waiting for @annaola review

File: izp/polls/tests/test_models.py [hunk 30]
```diff
@@ -137,6 +138,26 @@ def test_creating_empty_open_question(self):
...
+        self.assertIs(
+            'Odp1' in people_question and 'Odp2' in people_question, True)
+
+    def test_creating_empty_people_question(self):
+        poll = Poll.objects.create()
+        people_question = PeopleQuestion.objects.create(
+            poll=poll, question_text="PeopleQuestion")
+        self.assertIs(len(people_question.choice_set.all()), 0)
+
+
```
Comment (czapiga) [inline izp/polls/tests/test_models.py]: These tests are not testing anything interesting - anything specific to PeopleQuestion class. In these tests you're using PeopleQuestion object just like OpenQuestion one. You can change class name and test will still work - I think that's bad sign. Please remove these tests.

File: izp/polls/tests/test_models.py [hunk 31]
```diff
@@ -137,6 +138,26 @@ def test_creating_empty_open_question(self):
         self.assertIs(len(open_question.choice_set.all()), 0)
```
Comment (czapiga) [inline izp/polls/tests/test_models.py]: Where are tests for Employers class you've created?

File: izp/polls/tests/test_views.py [hunk 32]
```diff
@@ -123,6 +124,37 @@ def test_open_question_without_choices(self):
...
+    def test_people_question_without_choices(self):
+        '''
+        Test for detail view of empty people question
+        '''
+        people_question = PeopleQuestion.objects.get(
+            question_text="PeopleQuestion")
+        url = reverse('polls:question_detail', args=(people_question.id,))
+        response = self.client.get(url)
+        self.assertContains(response, people_question.question_text)
+        self.assertContains(response, 'new_choice')
```
Comment (czapiga) [inline izp/polls/tests/test_views.py]: The same as before. You're testing OpenQuestion class not PeopleQuestion. Where you're using fact that PeopleQuestion has list of employers in it's view? Remove these tests and replace them with something useful.

File: izp/polls/tests/test_views.py [hunk 33]
```diff
@@ -185,6 +217,48 @@ def test_no_answers_for_open_question(self):
             self, response, open_question, "Nie wybrano odpowiedzi")
 
 
+class PeopleQuestionVoteViewTests(TestCase):
```
Comment (czapiga) [inline izp/polls/tests/test_views.py]: Again.. nothing useful.

File: izp/polls/tests/test_views.py (hunk 32 above)
Comment (iCarrrot) [reply izp/polls/tests/test_views.py]: done

File: izp/polls/tests/test_views.py [hunk 34]
```diff
@@ -123,6 +121,42 @@ def test_open_question_without_choices(self):
...
+        people_question = PeopleQuestion.objects.get(
+            question_text="PeopleQuestion")
+        url = reverse('polls:question_detail', args=(people_question.id,))
+        response = self.client.get(url)
+        self.assertContains(response, people_question.question_text)
+        self.assertContains(response, 'new_choice')
+        self.assertContains(response, 'list="employers"')
+        self.assertContains(response, 'datalist id="employers"')
+
+    def test_people_question_without_choices(self):
```
Comment (annaola) [inline izp/polls/tests/test_views.py]: I think this test is the same as test_people_question_with_choices, as @czapiga has already mentioned, isn't it?

File: izp/polls/employers.py [hunk 35]
```diff
@@ -0,0 +1,34 @@
...
+
+class Employers:
+    """Creating list of employers from file
+    """
+
+    __list = None
+
+    @classmethod
+    def get_list(cls):
+        """Tak file polls/static/data/employers and convert
```
Comment (czapiga) [inline izp/polls/employers.py]: Typo

File: izp/polls/tests/test_models.py [hunk 36]
```diff
@@ -193,3 +197,21 @@ def test_adding_open_question(self):
             poll=poll, question_text="test-question")
         question_names = map(str, poll.question_set.all())
         self.assertIn(str(question), question_names)
+
+
+class EmployersClassTest(TestCase):
+    """Tests for Employers class
```
Comment (czapiga) [inline izp/polls/tests/test_models.py]: Inconsistent formatting

File: izp/polls/views.py [hunk 37]
```diff
@@ -148,10 +153,13 @@ def login(request, poll_id):
...
     question = get_object_or_404(Question, pk=question_id)
     is_open = OpenQuestion.objects.filter(pk=question.pk).exists()
+    is_people_question = PeopleQuestion.objects.filter(pk=question.pk).exists()
     is_session = 'poll' + str(question.poll.id) in request.session
-
+    employers_list = Employers.get_list()
     context = {'question': question,
                'is_open': is_open,
+               'is_people_question': is_people_question,
+               'employers': employers_list,
```
Comment (czapiga) [inline izp/polls/views.py]:
```python
'employers': Employers.get_list(),
```

File: izp/polls/views.py [hunk 38]
```diff
@@ -47,6 +51,7 @@ def question_detail(request, question_id):
 
     if not question.is_active():
         context['error'] = "Głosowanie nie jest aktywne"
+
```
Comment (czapiga) [inline izp/polls/views.py]: revert this change
